WiP: Other new sgde methods

spc90 opened the pull request:
This branch introduces another two new density estimation methods in the DM pipeline:

1. Sparse grid density derivative estimation (SGDDerivE): given a dataset representing samples from a distribution, we approximate the pdf of the 1st order partial derivative distribution along a given dimension. This is a direct extension of the SGDE method, i.e. the resulting system matrix grows only with grid points. Requires B-splines.

2. Sparse grid density derivative ratio estimation (SGDDerivRE): given a dataset representing samples from a distribution, we directly approximate the pdf of thedensity derivative ratio distribution. This is an extension of the LS method (regression), i.e. the resulting system matrix grows with data and grid points. Requires B-splines.

leiterrl commented on base/src/sgpp/base/operation/hash/OperationMultipleEvalPartialDerivativeModBsplineNaive.cpp:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+double OperationMultipleEvalPartialDerivativeModBsplineNaive::getDuration() { return 0.0; }
```
What's the purpose of this function and why return zero? Do other operations handle it like this as well?

spc90 replied:
All of the existing MultipleEval classes have that. I haven't quite looked into the purpose of it, I just wanted to have the same unified handling for this operation as well. Maybe the avx/ocl etc. use an actual value there. Not sure of the original purpose, so we might open a separate issue on that.

leiterrl commented on datadriven/src/sgpp/datadriven/algorithm/DBMatOnlineDE.cpp:
```diff
@@ -493,6 +569,82 @@ std::vector<DataVector> DBMatOnlineDE::computeWeightedBFromBatchTwoDatasets(
+          "In DBMatOnlineDE::computeWeightedBFromBatch: b doesn't match size of system matrix");
+    }
+
+    /*
+    std::unique_ptr<sgpp::base::OperationMultipleEval> B(
+        (offlineObject.interactions.size() == 0)
+            ? sgpp::op_factory::createOperationMultipleEval(grid, m)
+            : sgpp::op_factory::createOperationMultipleEvalInter(grid, m,
+                                                                 offlineObject.interactions));
+    */
```
There is a bunch of commented code in this file. This should be removed if it doesn't have a purpose anymore

spc90 replied:
All the commented pieces of code (also the ones you've marked below) serve as basis for extending these new approaches with features that exist for the regular density estimation. Things like geometry aware grids (what is here, this MultipleEvalInter), enabling the visualization module, enabling hyperparameter optimization etc. 
I wanted this PR to be the basic implementation of the algorithm, and then in a next PR I will go through all these features and see which of them make sense to be implemented for each new method or not. And it's easier and safer to have them commented out first and then I can remove things, so that I don't forget something :) If you think these commented lines will increase the git history too much, I can start with adding the features now in this pull request instead of a separate one that weeds them out.

leiterrl commented on datadriven/src/sgpp/datadriven/algorithm/DBMatOnlineDE.cpp:
```diff
@@ -565,20 +717,22 @@ DataVectorDistributed DBMatOnlineDE::computeWeightedBFromBatchParallel(
 
+/*
 double DBMatOnlineDE::resDensity(DataVector& alpha, Grid& grid) {
   auto C = sgpp::op_factory::createOperationIdentity(grid);
   DataVector rhs(grid.getSize());
   DataVector res(grid.getSize());
-  sgpp::datadriven::DensitySystemMatrix SMatrix(grid, *testMat, C, 0.0);
 
+  sgpp::datadriven::DensitySystemMatrix SMatrix(grid, *testMat, C, 0.0);
   SMatrix.generateb(rhs);
-
   SMatrix.mult(alpha, res);
 
   for (size_t i = 0; i < res.getSize(); i++) res[i] -= rhs[i];
   return res.l2Norm();
 }
+*/
 
+/*
```
Again the commented code should be removed

leiterrl commented on datadriven/src/sgpp/datadriven/datamining/builder/DensityDerivativeEstimationMinerFactory.cpp:
```diff
@@ -0,0 +1,92 @@
+  throw base::application_exception("Unknown density estimation type");
+}
+
+/*
+HyperparameterOptimizer *DensityDerivativeEstimationMinerFactory::buildHPO(
+    const std::string &path) const {
+  DataMiningConfigParser parser(path);
+  if (parser.getHPOMethod("bayesian") == "harmonica") {
+    return new HarmonicaHyperparameterOptimizer(buildMiner(path),
+                                                new DensityEstimationFitterFactory(parser), parser);
+  } else {
+    return new BoHyperparameterOptimizer(buildMiner(path),
+                                         new DensityEstimationFitterFactory(parser), parser);
+  }
+}
+*/
+
+/*
+FitterFactory *DensityDerivativeEstimationMinerFactory::createFitterFactory(
+    const DataMiningConfigParser &parser) const {
+  return new DensityEstimationFitterFactory(parser);
+}
+*/
```
Commented code again

leiterrl commented on datadriven/src/sgpp/datadriven/datamining/builder/DensityDerivativeRatioEstimationMinerFactory.cpp:
```diff
@@ -0,0 +1,58 @@
+        "order at least 3!");
+}
+
+/*
+FitterFactory* DensityDerivativeRatioEstimationMinerFactory::createFitterFactory(
+    const DataMiningConfigParser& parser) const {
+  return new LeastSquaresRegressionFitterFactory(parser);
+}
+*/
```
And again commented code

leiterrl commented on datadriven/src/sgpp/datadriven/datamining/builder/RelativeDensityRatioEstimationMinerFactory.cpp:
```diff
@@ -0,0 +1,108 @@
+  return dataSources;
+}
+
+/*
+ std::vector<DataSourceCrossValidation*>
+ RelativeDensityRatioEstimationMinerFactory::createDataSourceCrossValidationTwoDatasets(
+ const DataMiningConfigParser& parser) const {
+ DataSourceConfig config { };
+
+ CrossvalidationConfiguration crossValidationconfig { };
+ parser.getFitterCrossvalidationConfig(crossValidationconfig,
+ crossValidationconfig);
+
+ bool hasSource = parser.getDataSourceConfig(config, config);
+
+ if (hasSource && config.filePath.compare("") != 0) {
+ DataSourceBuilder builder;
+ return builder.crossValidationFromConfig(config, crossValidationconfig);
+ } else {
+ throw base::data_exception("No file name provided for datasource.");
+ }
+ }
+ */
```
And again

leiterrl commented on datadriven/src/sgpp/datadriven/datamining/builder/RelativeDensityRatioEstimationMinerFactory.cpp:
```diff
@@ -0,0 +1,108 @@
+  return new ModelFittingRelativeDensityRatioEstimation(config);
+}
+
+/*
+ FitterFactory* RelativeDensityRatioEstimationMinerFactory::createFitterFactory(
+ const DataMiningConfigParser& parser) const {
+ return new DensityRatioEstimationMinerFactory(parser);
+ }
+ */
```
And again :-)

spc90 commented on pde/src/sgpp/pde/operation/hash/OperationMatrixLTwoDotExplicitBspline.hpp:
```diff
@@ -48,6 +56,104 @@ class OperationMatrixLTwoDotExplicitBspline : public sgpp::base::OperationMatrix
+   * @param j_start start index for column iteration
+   * @param j_end end index for column iteration
+   */
+  inline void buildMatrixWithBounds(sgpp::base::DataMatrix* mat, sgpp::base::Grid* grid,
```
I might have gone with the flow on this one. The linear and modlinear versions (which I based on) are technically less complex, but the main purpose, I guess also for those, was to remove some of this unnecessary context switching happening, as these functions are always called in other small builder functions that basically don't do anything else than passing different specific values for those bounds. Doing a proper performance analysis to see the impact of this is I guess not that easy, and the compiler is anyway not forced to inline if it thinks it shouldn't... but I'm happy to change all of them if you think this is pointless or hurts the performance.

spc90 commented on pde/src/sgpp/pde/operation/hash/OperationMatrixLTwoDotExplicitModBspline.hpp:
```diff
@@ -48,6 +57,107 @@ class OperationMatrixLTwoDotExplicitModBspline : public sgpp::base::OperationMat
+   * @param j_start start index for column iteration
+   * @param j_end end index for column iteration
+   */
+  inline void buildMatrixWithBounds(sgpp::base::DataMatrix* mat, sgpp::base::Grid* grid,
```
See above

spc90 commented on base/examples/benchmark_gridInteraction.cpp:
```diff
@@ -36,9 +36,9 @@ void genAllInteractions(std::set<std::set<size_t>> &inter, size_t d) {
-  std::vector<size_t> dims = {1, 2, 3};
-  size_t lvl_start = 3;
-  size_t lvl_end = 5;
+  std::vector<size_t> dims = {1, 2};
+  size_t lvl_start = 2;
+  size_t lvl_end = 4;
```
It just shortens the overall time of the datadriven tests, as it doesn't really need to go through all those costlier combinations :) Not directly related to this PR though, indeed...

spc90 commented:
I have finally revived this pull request 🥳, so that I can merge these methods from my thesis. Once the PR goes through, I will also double-check and update the wiki to reflect the current state of the DM pipeline.

@leiterrl I have addressed the main concerns you had back then, when you first looked at this. Since then, I have added some of the functionalities which were commented out, and the rest of the commented sections were removed, so that got cleaned. Regarding the OpFactory methods, those are needed because these density methods require Bspline bases (higher degree bases in general), and these operations that I needed were not implemented. It probably could have gone in a separate, earlier PR, but I did the implementation in sort-of reverse order (first the methods, then the operations) which was not ideal... 🙄 

When you have time, please have a fresh look at this PR and let me know if there are still issues you see that I should address.

Also, there seems to be a problem with the Travis jobs on Windows. It looks like the VM is failing to install packages before actually building the code. Can you look into it, please? I seem to recall similar issues in the past... 🤔 Maybe just restarting the jobs would work.
